### ISSUE TITLE
REST API multiselect attribute BC break between 2.1.8 and 2.3.0

### ISSUE THREAD

**Issue by mhujer** [magento/magento2#21364]
### Preconditions (*)

1. Magento 2.4-develop

### Steps to reproduce (*)

1. Have a multiselect attribute
2. Update the product data through the REST API and pass an array of IDs as a value for this multiselect attribute

### Expected result (*)

1. Attribute is updated (as it worked in 2.1.8)

### Actual result (*)

1. Error message is returned:

```
{
    "message": "Error occurred during \"custom_attributes\" processing. Attribute \"equipment\" has invalid value. The \"array\" value's type is invalid. The \"string\" type was expected. Verify and try again.",
    "trace": null
}
```

### Workaround
Implode the ids and send them as a string. But it makes your code incompatible with Magento 2.1.8 as it expects an array of ids.

I guess that it is some kind of BC break, because Magento 2.1.8 works in a completely opposite way - it requires an array of IDs and rejects string with a message: `{"message":"Error occurred during \"custom_attributes\" processing. Attribute \"equipment\" has invalid value. Invalid type for value: \"string\". Expected Type: \"string[]\"."}`

Related issue: #6120

**Comment by magento-engcom-team** [comment]
Hi @mhujer. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me 2.3-develop instance` - upcoming 2.3.x release

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@mhujer do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by ankit-sf** [comment]
Now  _array of ids_ values is changed by _comma seperated ids_

**Comment by mhujer** [comment]
@ankit-sf So this change was intentional and is expected?

**Comment by ankit-sf** [comment]
no, i **don't** think it is intentional, it's just that it is not corrected in 2.3 but is corrected in previous versions 2.2 and 2.1

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

-  _Next steps are available in case you are a member of **Community Maintainers**._

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-backlog-nazar 
 Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-98363`, `MAGETWO-98364` were created

**Issue Available**: @engcom-backlog-nazar, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by magento-engcom-team** [comment]
Hi @Valant13. Thank you for working on this issue. 
Looks like this issue is already verified and confirmed. But if you want to validate it one more time, please, go though the following instruction:
- [ ] 1. Add/Edit `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.

- [ ] 2. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 3. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 4. If the issue is not relevant or is not reproducible any more, feel free to close it.

**Comment by onepack** [comment]
Any progress on this? The last update was in May. It's October now.

**Comment by zap263** [comment]
> Any progress on this? The last update was in May. It's October now.

See post by @unicoder88

**Comment by onepack** [comment]
Is this tested with just only admin values applied?
When I add translations it seems that the multiselect breaks again.

**Comment by lano-vargas** [comment]
This is still an issue on 2.4, has not got any updates for this here:https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Eav/Model/TypeLocator/SimpleType.php

Is this the new way for sending setting multiselect attribute can someone comfirm?

Thanks

**Comment by LiamKarlMitchell** [comment]
Weirdly having this problem in a live site, but on the staging site which was a copy of the live site at one point (same Magento 2.3.0 version) it does not have the problem.
The environments are identical as far as I can tell.

One Magento 2.3.0

Attribute "producttype" has invalid value. The "array" value's type is invalid. The "string" type was expected. Verify and try again.

**Comment by LiamKarlMitchell** [comment]
Confirming that the patch https://github.com/ConvertGroupsAS/magento2-patches/commit/fde2602cc3ceed9892971406b2b48b903fc60fd7 fixes this.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Charlie 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-24135` were created

**Issue Available**: @engcom-Charlie, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by engcom-Delta** [comment]
Hi @mhujer ,

Thanks for your reporting and collaboration.
We have verified the issue in latest 2.4-develop instance and the issue is reproducible.
Kindly refer the screenshots.

**Steps to reproduce**
1. Have a multiselect attribute
2. Update the product data through the REST API and pass an array of IDs as a value for this multiselect attribute
Error displays in response, instead of updating attribute value - 
![image](https://github.com/user-attachments/assets/b7e0b49c-3dbb-4e8c-bf3d-106f7e57eebc)

Thanks.

**Comment by kestraly** [comment]
The patch still works perfectly - Setups like Linnworks require this to get their mutli select attributes to work

https://github.com/ConvertGroupsAS/magento2-patches/commit/fde2602cc3ceed9892971406b2b48b903fc60fd7

**Comment by lbajsarowicz** [comment]
engcom-Delta reconfirmed on 2024-09-25 that multiselect custom_attributes reject an array, a behavior change since 2.3.0 versus 2.1.8. Two prior PRs (#21901, #31930) were closed unmerged, and Kestraly reported in 2025-05 that a private patch still works. This needs a maintainer decision on the correct array/string contract before another PR; please re-QA and confirm which contract is intended.
